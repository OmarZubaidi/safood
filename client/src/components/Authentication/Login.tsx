// Package imports
import React, { useState, useRef } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Card, Button, Form, Alert } from 'react-bootstrap';

// Local imports
import { useAuth } from '../../context/AuthContext';
// import { LoginUser } from '../../interfaces/Authentication/Login.interface';

export default function Login () {
  // Refs, states, and navigation
  const emailRef = useRef<HTMLInputElement>();
  const passwordRef = useRef<HTMLInputElement>();
  const { login } = useAuth();
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>(false);
  const navigate = useNavigate();

  // Submit function
  async function handleSubmit (e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    try {
      setError('');
      setLoading(true);

      await login(
        emailRef.current.value,
        passwordRef.current.value
      );
      navigate('/');
    } catch (error) {
      console.error(error);
      setError('Failed to sign in.');
    }
    setLoading(false);
  }

  return (
    <>
      <Card>
        <Card.Body>
          <h2 className='text-center mb-4'>
            Log in
          </h2>
          {error && <Alert variant='danger'>
            {error}
          </Alert>}
          <Form onSubmit={handleSubmit}>
            <Form.Group id='email'>
              <Form.Label>
                E-mail
              </Form.Label>
              <Form.Control
                type='email'
                ref={emailRef}
                required
              />
            </Form.Group>
            <Form.Group id='password'>
              <Form.Label>
                Password
              </Form.Label>
              <Form.Control
                type='password'
                ref={passwordRef}
                required
              />
            </Form.Group>
            <Button
              disabled={loading}
              className='w-100 mt-4'
              type='submit'
            >
              Log in
            </Button>
          </Form>
        </Card.Body>
      </Card>
      <div className='w-100 text-center mt-2'>
        Need an account? <Link to='/signup'>
          Sign up
        </Link>
      </div>
    </>
  );
}
