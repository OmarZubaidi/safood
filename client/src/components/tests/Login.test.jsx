// Package imports
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Local imports
import Login from '../Authentication/Login';

jest.mock('../../context/AuthContext', () => ({
  useAuth: () => {
    return {
      login: (mockLogin) => {
        // console.log(mockLogin);
      }
    };
  }
}));

describe('Login component', () => {

  test('should match the snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render the login component', () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    );
    screen.getByText('E-mail');
    screen.getByText('Password');
    screen.getByRole('button');
  });


});

