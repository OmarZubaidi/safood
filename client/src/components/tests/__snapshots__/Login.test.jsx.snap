// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should match the snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h2
        class="text-center mb-4"
      >
        Log in
      </h2>
      <form
        class=""
      >
        <div
          id="email"
        >
          <label
            class="form-label"
            for="email"
          >
            E-mail
          </label>
          <input
            aria-label="email"
            class="form-control"
            name="email"
            required=""
            type="email"
          />
        </div>
        <div
          id="password"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            aria-label="password"
            class="form-control"
            name="password"
            required=""
            type="password"
          />
        </div>
        <button
          class="w-100 mt-4 btn btn-primary"
          type="submit"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
  <div
    class="w-100 text-center mt-2"
  >
    Need an account? 
    <a
      href="/signup"
    >
      Sign up
    </a>
  </div>
</div>
`;
